src/actions/attack.cpp

```diff
@@ -143,9 +143,9 @@ battle_context_unit_stats::battle_context_unit_stats(const unit& u,
 	petrifies = weapon->bool_ability("petrifies");
 	poisons = !opp.get_state("unpoisonable") && weapon->bool_ability("poison") && !opp.get_state(unit::STATE_POISONED);
 	backstab_pos = is_attacker && backstab_check(u_loc, opp_loc, units, resources::gameboard->teams());
-	rounds = weapon->get_specials("berserk").highest("value", 1).first;
+	rounds = std::max(0, weapon->get_specials("berserk").highest("value", 1).first);
 	if(weapon->combat_ability("berserk", 1).second) {
-		rounds = weapon->combat_ability("berserk", 1).first;
+		rounds = std::max(0, weapon->combat_ability("berserk", 1).first);
```

Comment:
Repeating my comment from #3918: I'm not sure this is a bug. Using -1 to represent "no limit" is not unheard of. It's at least as defensible as interpreting -1 as an upper bound on the number of rounds.

I think this might break `[berserk] value=-1 cumulative=yes [/berserk]` but haven't tested.

All in all, I am leaning against this pull request. It's likely to break more things than it's likely to fix.